wagtail/users/views/users.py

```diff
@@ -140,44 +143,42 @@ def index(request, *args):
         )
 
 
-@permission_required(add_user_perm)
-def create(request):
-    for fn in hooks.get_hooks("before_create_user"):
-        result = fn(request)
-        if hasattr(result, "status_code"):
-            return result
-    if request.method == "POST":
-        form = get_user_creation_form()(request.POST, request.FILES)
-        if form.is_valid():
-            with transaction.atomic():
-                user = form.save()
-                log(user, "wagtail.create")
-            messages.success(
-                request,
-                _("User '{0}' created.").format(user),
-                buttons=[
-                    messages.button(
-                        reverse("wagtailusers_users:edit", args=(user.pk,)), _("Edit")
-                    )
-                ],
-            )
-            for fn in hooks.get_hooks("after_create_user"):
-                result = fn(request, user)
-                if hasattr(result, "status_code"):
-                    return result
-            return redirect("wagtailusers_users:index")
-        else:
-            messages.error(request, _("The user could not be created due to errors."))
-    else:
-        form = get_user_creation_form()()
+class Create(CreateView):
 
-    return TemplateResponse(
-        request,
-        "wagtailusers/users/create.html",
-        {
-            "form": form,
-        },
-    )
+    permission_policy = ModelPermissionPolicy(User)
+    permission_required = "add"
+    form_class = get_user_creation_form()
+    template_name = "wagtailusers/users/create.html"
+    success_url = "wagtailusers_users:index"
+    add_url_name = "wagtailusers_users:add"
+    index_url_name = "wagtailusers_users:index"
+    page_title = gettext_lazy("Add user")
+
+    def run_before_hook(self):
+        for fn in hooks.get_hooks("before_create_user"):
+            result = fn(self.request)
+            if hasattr(result, "status_code"):
+                return result
+
+    def run_after_hook(self):
+        for fn in hooks.get_hooks("after_create_user"):
+            result = fn(self.request, self.object)
+            if hasattr(result, "status_code"):
+                return result
+        return redirect("wagtailusers_users:index")
+
+    def get_success_message(self, instance):
+        return _("User '{0}' created.").format(instance)
+
+    def get_success_buttons(self):
+        return (
+            [
+                messages.button(
+                    reverse("wagtailusers_users:edit", args=(self.object.pk,)),
+                    _("Edit"),
+                )
+            ],
+        )
```

Comment:
This can be a lot simpler - the `CreateView` has a few more helpful methods and attributes to save you time. It's good to take a look at other code that uses this kind of thing, especially the hooks.

Also, pulling out the URLs to the methods helps to simplify things, I had to re-add `get_add_url` to return `None` so that the add page would not show an empty 'add' button at the top right.

```suggestion
    """
    Provide the ability to create a user within the admin.
    """

    permission_policy = ModelPermissionPolicy(User)
    permission_required = "add"
    form_class = get_user_creation_form()
    template_name = "wagtailusers/users/create.html"
    add_url_name = "wagtailusers_users:add"
    index_url_name = "wagtailusers_users:index"
    edit_url_name = "wagtailusers_users:edit"
    success_message = "User '{0}' created."
    page_title = gettext_lazy("Add user")

    def run_before_hook(self):
        return self.run_hook(
            "before_create_user",
            self.request,
        )

    def run_after_hook(self):
        return self.run_hook(
            "after_create_user",
            self.request,
            self.object,
        )

    def get_add_url(self):
        return None
```

Reply:
Ahh I see why you re-wrote that `get_success_buttons`

I will take a look at your object.pk/id thing next but re-add the method override here for now.